프로필 상세 Phase 2

quessr opened the pull request:
## 💡 관련 이슈
#20 

## 🌱 변경 사항 및 이유
- header, footer 레이아웃 추가
- 스크롤 위치에 따라 헤더색상 변경 기능 구현
- 닫기버튼 클릭시 메인화면으로 이동
- 좋아요 토글 기능 구현
- 동영상 미디어 연결 및 사운드 아이콘 색상변경 적용

## ❗️ 참고 사항

quessr committed:
CoordinatorLayout 및 AppBarLayout을 추가하여 헤더 레이아웃 구현 #20

quessr committed:
스크롤 위치에 따라 헤더 색상을 변경하는 RecyclerView 스크롤 리스너 추가 #20

quessr committed:
footer 레이아웃 구현 #20

quessr committed:
구성 변경 및 스크롤 시 ViewPager2 위치 유지 구현 #20

quessr committed:
닫기 버튼 클릭 시 메인 화면으로 이동 기능 추가 #20

quessr committed:
오버플로 메뉴 클릭시 신고하기/차단하기 다이얼로그 연결 추가 #20

quessr committed:
좋아요 토글 기능 구현 #20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

# Conflicts:
#	app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailFragment.kt

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailViewModel.kt:
```diff
@@ -94,4 +108,20 @@ class ProfileDetailViewModel : ViewModel() {
+
+    fun toggleLike() {
+        val currentItems = _items.value.toMutableList()
+        val profileImagesIndex = currentItems.indexOfFirst { it is ProfileDetailModel.ProfileImages }
+
+        if (profileImagesIndex != -1) {
+            val profileImages = currentItems[profileImagesIndex] as ProfileDetailModel.ProfileImages
```
currentItems.filterIsInstance<ProfileDetailModel.ProfileImages>().firstOrNull()로 줄일 수 있습니다~(nullable)

quessr replied:
> currentItems.filterIsInstance<ProfileDetailModel.ProfileImages>().firstOrNull()로 줄일 수 있습니다~(nullable)

알려주셔서 감사합니다 너무 마음에 듭니다 ㅎㅎ 최고 👍

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailAdapter.kt:
```diff
@@ -37,7 +37,13 @@ class ProfileDetailAdapter(
+                ProfileDetailViewHolder.Images(
+                    binding,
+                    onPageChangeCallback = { currentPosition ->
+                        viewModel.setCurrentPosition(currentPosition)
```
adapter에 viewModel을 직접 주입하시기보단
람다식이나 인터페이스 형태로 주입하시는걸 권장드립니다!
특정 ViewModel에 종속성이 생기면 재사용이나 확장(+상속)이 힘들어져서요

quessr replied:
> adapter에 viewModel을 직접 주입하시기보단 람다식이나 인터페이스 형태로 주입하시는걸 권장드립니다! 특정 ViewModel에 종속성이 생기면 재사용이나 확장(+상속)이 힘들어져서요

람다식으로 변경했습니다~ 알려주셔서 감사합니다 도움이 많이 되었습니다 🙏

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailFragment.kt:
```diff
@@ -28,12 +36,40 @@ class ProfileDetailFragment() :
         setupInitialData()
     }
 
+    override fun onViewStateRestored(savedInstanceState: Bundle?) {
+        super.onViewStateRestored(savedInstanceState)
+
+        // ViewModel의 currentPosition 값을 복원
+        profileDetailViewModel.restoreCurrentPosition(
+            savedInstanceState?.getInt(
+                CURRENT_POSITION_KEY
+            )
+        )
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        // ViewModel의 currentPosition 값을 저장
+        outState.putInt(CURRENT_POSITION_KEY, profileDetailViewModel.currentPosition.value)
+    }
+
```
saveInstance에 따로 저장하실 필요는 없을 것 같아요!
page가 바뀔 때 마다 이미 뷰모델에 저장하고 있기 때문에 아마 저장된 값과 같은 값을 저장할거에요

quessr replied:
> saveInstance에 따로 저장하실 필요는 없을 것 같아요! page가 바뀔 때 마다 이미 뷰모델에 저장하고 있기 때문에 아마 저장된 값과 같은 값을 저장할거에요

네 이부분도 확인해 보니까 필요없어서 삭제 했습니다 감사합니다 🙏

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailViewModel.kt:
```diff
@@ -15,6 +16,11 @@ class ProfileDetailViewModel : ViewModel() {
     private val _qnaItems = MutableStateFlow<List<ProfileDetailModel.Qna>>(emptyList())
     val qnaItems: StateFlow<List<ProfileDetailModel.Qna>> get() = _qnaItems
 
+    private val _currentPosition = MutableStateFlow(0)
+    val currentPosition: StateFlow<Int> get() = _currentPosition
```
관측하는 값이 아니라 단순히 set get하는 값들은 그냥 변수로 만드시는게 어떨까요??

quessr replied:
> 관측하는 값이 아니라 단순히 set get하는 값들은 그냥 변수로 만드시는게 어떨까요??

이 부분도 제가 무지성으로 사용하고 있었네요 😅 짚어주셔서 감사합니다 🙏

quessr committed:
동영상 미디어 처리 및 뮤트 아이콘 가시성 업데이트 기능 추가 #20

quessr committed:
ViewPagerAdapter에 동영상 미디어 지원 추가 및 적용 #20

- 동영상과 이미지를 구분하여 처리하도록 ViewHolder 구조 변경
- ImageViewHolder와 VideoViewHolder를 통해 각각의 미디어 타입 처리
- 동영상 미디어의 경우 MediaView를 사용하여 썸네일과 재생 기능 추가

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

quessr committed:
중복된 ic drawable 파일 제거 및 기존 파일로 레이아웃 리소스 적용 #20

quessr committed:
footer 투표하기(좋아요) 버튼 디자인 변경에 따른 스타일 및 텍스트 업데이트 #20

quessr committed:
Footer와 Images의 좋아요 상태 동기화 처리 #20

- ProfileDetailViewModel의 isLike 상태 관리 추가
- 신고하기/차단하기 로직 주석 처리

quessr committed:
볼륨 토글 기능 추가 및 스크롤에 따른 사운드 아이콘 색상 변경 적용 #20

- MediaDetailViewModel의 isVolumeOn 상태를 활용해 사운드 아이콘 업데이트

quessr committed:
toggleLike 함수 리팩토링 #20

- ProfileImages 인스턴스를 필터링하여 처리
- 코드 정리

quessr committed:
currentPosition StateFlow에서 일반 변수로 변경 #20

quessr committed:
ViewModel 종속성을 제거하고 람다식으로 이미지 포지션 및 좋아요 상태 변경 처리 수정 #20

- onPageChangeCallback 함수명을 구체적으로 변경

quessr committed:
필요없는 코드 삭제 #20

quessr committed:
불필요한 MutableStateFlow 제거 및 일반 변수를 사용하도록 수정 #20

quessr commented:
> ## 💡 관련 이슈
> #20
> 
> ## 🌱 변경 사항 및 이유
> * header, footer 레이아웃 추가
> * 스크롤 위치에 따라 헤더색상 변경 기능 구현
> * 닫기버튼 클릭시 메인화면으로 이동
> * 좋아요 토글 기능 구현
> * 신고하기/차단하기 다이얼로그 연결
> * viewpager2 position 유지 기능 구현
> * 동영상 미디어 연결 및 사운드 아이콘 색상변경 적용
> 
> ## ❗️ 참고 사항

@kim0hoon 
영훈님 피드백 전부 반영 완료했습니다~
시간나실때 확인 해 보시고 더 수정 해야 될 사항 있으면 말씀 주시면 감사하겠습니다 😊

quessr committed:
viewLifecycleOwner에 repeatOnLifecycle 추가 및 findNavController 사용하여 네비게이션 코드 간소화 #20

quessr committed:
ViewModel 종속성 제거 #20

QnaToggle 및 currentPosition 전달 로직 추가

quessr committed:
currentPosition 전달을 위한 콜백 추가 및 QnaToggle 처리 로직 수정 #20

quessr committed:
주석 삭제 #20

quessr committed:
회원 상세 정보 보기 API 연결 기능 구현 #20

- UserRepository에 fetchUserDetailInfo 메서드 추가
- GetUserDetailInfoUseCase 추가
- UserService 인터페이스에 /v1/api/user/find/{uniqueId} API 연결
- ProfileDetailResponse 및 ProfileDetailMapper 추가
- ProfileDetailViewModel에서 사용자 프로필 정보를 가져올 때 GetUserDetailInfoUseCase를 사용하도록 변경
- ModelState 및 Job 적용

quessr committed:
close 버튼 클릭시 네비게이션 동작 수정 #20

- 화면 전환에서 이전 화면으로 돌아가기 처리로 수정

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

quessr committed:
신고하기/차단하기 다이얼로그 추가 #20

quessr committed:
QnA 리스트가 없는 경우 UI 처리 #20

quessr committed:
주석 삭제 #20

quessr committed:
fetchUserDetailInfo 메소드에 더미 데이터 추가 #20

- 미디어 링크를 비디오와 이미지로 구분하는 로직 추가

quessr committed:
초기 데이터 설정 로직 제거 #20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

quessr committed:
채팅 메뉴 리소스 ID 수정 #20

quessr committed:
신고하기/차단하기 다이얼로그 적용 #20
